% Template created by Karol Kozio≈Ç (www.karol-koziol.net) for ShareLaTeX

\documentclass[a4paper,9pt]{extarticle}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{xcolor}
\usepackage{tikz}

\usepackage{amsmath,amssymb,textcomp}
\everymath{\displaystyle}

\usepackage{times}
\renewcommand\familydefault{\sfdefault}
\usepackage{tgheros}
\usepackage[defaultmono,scale=0.85]{droidmono}

\usepackage{multicol}
\setlength{\columnseprule}{0pt}
\setlength{\columnsep}{20.0pt}


\usepackage{geometry}
\geometry{
a4paper,
total={210mm,297mm},
left=10mm,right=10mm,top=10mm,bottom=15mm}

\linespread{1.3}


% custom title
\makeatletter
\renewcommand*{\maketitle}{%
\noindent
\begin{minipage}{0.65\textwidth}
\begin{tikzpicture}
\node[rectangle,rounded corners=6pt,inner sep=10pt,fill=blue!50!black,text width= 0.95\textwidth] {\color{white}\Huge \@title};
\end{tikzpicture}
\end{minipage}
\hfill

\bigskip\bigskip
}%
\makeatother

% custom section
\usepackage[explicit]{titlesec}
\newcommand*\sectionlabel{}
\titleformat{\section}
  {\gdef\sectionlabel{}
   \normalfont\sffamily\Large\bfseries\scshape}
  {\gdef\sectionlabel{\thesection\ }}{0pt}
  {
\noindent
\begin{tikzpicture}
\node[rectangle,rounded corners=3pt,inner sep=4pt,fill=blue!50!black,text width= 0.95\columnwidth] {\color{white}\sectionlabel#1};
\end{tikzpicture}
  }
\titlespacing*{\section}{0pt}{15pt}{10pt}


% custom footer
\usepackage{fancyhdr}
\makeatletter
\pagestyle{fancy}
\fancyhead{}
\fancyfoot[C]{\footnotesize \textcopyright\ \@date\ \ \@author}
\renewcommand{\headrulewidth}{0pt}
\renewcommand{\footrulewidth}{0pt}
\makeatother


\title{Cortex for bacterial genomics: quickstart}
\date{2015}



\begin{document}

\maketitle

\begin{multicols*}{2}


\section{Requirements}

Download Cortex from github (https://github.com/iqbal-lab/cortex)  - there are download instructions there.

You must have installed VCFtools (and have the entire directory, not just the binary executable), vcflib (from Erik Garrison, on github),  R and minimap2. R must be in your path. Add the following directories to PERL5LIB and PATH
\begin{verbatim}
export PERL5LIB= /path/cortex/scripts/analyse_variants/
    bioinf-perl/lib;/path/cortex/scripts/calling:
    /path/VCFTools_dir/perl:$PERL5LIB
export PATH = /path/cortex/scripts/analyse_variants/
   needleman_wunsch-0.3.0;/path/vcflib/bin:$PATH
\end{verbatim}
Also, make an INDEX file, mapping sample-identifiers to sequence data. 

\section{What is the Independent workflow?}
Do variant discovery independently for each sample (against a reference), then combine the callsets to make a single set of candidate sites (SNPs, indels, SVs), and then genotype all samples. 


This new pipeline will parallelise and scale to hundreds of thousands of bacteria. This cheatsheet shows how to run it on a single machine with many CPUs/cores, using GNU parallels.

\section{Independent workflow: how to run it}
This script will compile Cortex for you and make reference genome binaries - all the things that used to have to be done manually. It will also choose appropriate memory-use parameters
\begin{verbatim}
 perl cortex/scripts/calling/run_indep_wkflow_with_gnu_par.pl  
 --index INDEX  --ref_fa ref_genome.fa  
 --dir_for_ref_objects ref/ 
 --vcftools_dir ~/installed_apps/vcftools_0.1.9/ 
 --outdir results/ 
 --minimap2_bin ~/installed_apps/minimap2
 --kmer 31 --procs 20 --prefix salmonella
\end{verbatim}
This starts with FASTQ (or BAM) files, and finishes with a single VCF with all samples genotyped at the same sites. 

\section{Troubleshooting}
If some of your samples have extremely high coverage, it is possible the default memory use parameters won't be enough. 


\section{Segregating variants within our dataset (joint workflow)}
First build sample graphs as Step1 in the previous example. Then UNDETERMINED



\section{Pan-genome analysis}

To detect presence of a set of predefined genes (genes.fasta) among your samples
\begin{verbatim}
\end{verbatim}
To look at pan-genome graph of all samples and see which samples have which contigs, allowing you to stratify them by frequency or look for differentiating/segregating contigs.


\section{The End}

For further information:


\end{multicols*}

\end{document}
